### ISSUE TITLE
Heap Use-After-Free issue in chttp2 reported by fuzzer

### ISSUE THREAD

**Issue by yashykt** [grpc/grpc#14375]
```
Sanitizer report:
=================================================================
==498769==ERROR: AddressSanitizer: heap-use-after-free on address 0x629000008a20 at pc 0x5647193214e4 bp 0x7ffcff1965a0 sp 0x7ffcff196598
READ of size 8 at 0x629000008a20 thread T0
    #0 0x5647193214e3 in gpr_mpscq_pop_and_check_end(gpr_mpscq*, bool*) third_party/grpc/src/core/lib/gpr/mpscq.cc:49:43
    #1 0x564719321237 in gpr_mpscq_pop(gpr_mpscq*) third_party/grpc/src/core/lib/gpr/mpscq.cc:44:10
    #2 0x5647192d93b7 in grpc_combiner_continue_exec_ctx() third_party/grpc/src/core/lib/iomgr/combiner.cc:244:25
    #3 0x5647192e12e3 in grpc_core::ExecCtx::Flush() third_party/grpc/src/core/lib/iomgr/exec_ctx.cc:128:17
    #4 0x564718f2a89e in grpc_core::ExecCtx::~ExecCtx() third_party/grpc/src/core/lib/iomgr/exec_ctx.h:85:5
    #5 0x56471928e470 in grpc_server_shutdown_and_notify third_party/grpc/src/core/lib/surface/server.cc:1219:1
    #6 0x564718f24a39 in LLVMFuzzerTestOneInput third_party/grpc/test/core/end2end/fuzzers/api_fuzzer.cc:892:11
    #7 0x564719336823 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) third_party/llvm/llvm/projects/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:515:13
    #8 0x5647193255c4 in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) third_party/llvm/llvm/projects/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:280:6
    #9 0x56471932c271 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) third_party/llvm/llvm/projects/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:703:9
    #10 0x564719341d62 in main third_party/llvm/llvm/projects/compiler-rt/lib/fuzzer/FuzzerMain.cpp:20:10
    #11 0x7f3aff398bbc in __libc_start_main (/usr/grte/v4/lib64/libc.so.6+0x38bbc)
    #12 0x564718e525c8 in _start /usr/grte/v4/debug-src/glibc-2.19/csu/../sysdeps/x86_64/start.S:108

0x629000008a20 is located 14368 bytes inside of 16520-byte region [0x629000005200,0x629000009288)
freed by thread T0 here:
    #0 0x564718f0c442 in __interceptor_free third_party/llvm/llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:68:3
    #1 0x564719228d2b in destruct_transport(grpc_chttp2_transport*) third_party/grpc/src/core/ext/transport/chttp2/transport/chttp2_transport.cc:204:3
    #2 0x56471923edb7 in destroy_transport_locked(void*, grpc_error*) third_party/grpc/src/core/ext/transport/chttp2/transport/chttp2_transport.cc:572:3
    #3 0x5647192d9491 in grpc_combiner_continue_exec_ctx() third_party/grpc/src/core/lib/iomgr/combiner.cc:261:5
    #4 0x5647192e12e3 in grpc_core::ExecCtx::Flush() third_party/grpc/src/core/lib/iomgr/exec_ctx.cc:128:17
    #5 0x564718f2a89e in grpc_core::ExecCtx::~ExecCtx() third_party/grpc/src/core/lib/iomgr/exec_ctx.h:85:5
    #6 0x56471928e470 in grpc_server_shutdown_and_notify third_party/grpc/src/core/lib/surface/server.cc:1219:1
    #7 0x564718f24a39 in LLVMFuzzerTestOneInput third_party/grpc/test/core/end2end/fuzzers/api_fuzzer.cc:892:11
    #8 0x564719336823 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) third_party/llvm/llvm/projects/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:515:13
    #9 0x5647193255c4 in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) third_party/llvm/llvm/projects/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:280:6
    #10 0x56471932c271 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) third_party/llvm/llvm/projects/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:703:9
    #11 0x564719341d62 in main third_party/llvm/llvm/projects/compiler-rt/lib/fuzzer/FuzzerMain.cpp:20:10
    #12 0x7f3aff398bbc in __libc_start_main (/usr/grte/v4/lib64/libc.so.6+0x38bbc)
    #13 0x564718e525c8 in _start /usr/grte/v4/debug-src/glibc-2.19/csu/../sysdeps/x86_64/start.S:108

previously allocated by thread T0 here:
    #0 0x564718f0c997 in __interceptor_calloc third_party/llvm/llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:97:3
    #1 0x56471931aa15 in gpr_zalloc third_party/grpc/src/core/lib/gpr/alloc.cc:68:7
    #2 0x564719232985 in grpc_create_chttp2_transport(grpc_channel_args const*, grpc_endpoint*, bool) third_party/grpc/src/core/ext/transport/chttp2/transport/chttp2_transport.cc:3121:31
    #3 0x564718f97620 in on_handshake_done(void*, grpc_error*) third_party/grpc/src/core/ext/transport/chttp2/client/chttp2_connector.cc:117:9
    #4 0x5647192e15e3 in exec_ctx_run(grpc_closure*, grpc_error*) third_party/grpc/src/core/lib/iomgr/exec_ctx.cc:38:3
    #5 0x5647192e1370 in grpc_core::ExecCtx::Flush() third_party/grpc/src/core/lib/iomgr/exec_ctx.cc:125:9
    #6 0x564718f2a89e in grpc_core::ExecCtx::~ExecCtx() third_party/grpc/src/core/lib/iomgr/exec_ctx.h:85:5
    #7 0x5647192d0581 in grpc_timer_manager_tick() third_party/grpc/src/core/lib/iomgr/timer_manager.cc:104:1
    #8 0x564718f24647 in LLVMFuzzerTestOneInput third_party/grpc/test/core/end2end/fuzzers/api_fuzzer.cc:808:5
    #9 0x564719336823 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) third_party/llvm/llvm/projects/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:515:13
    #10 0x5647193255c4 in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) third_party/llvm/llvm/projects/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:280:6
    #11 0x56471932c271 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) third_party/llvm/llvm/projects/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:703:9
    #12 0x564719341d62 in main third_party/llvm/llvm/projects/compiler-rt/lib/fuzzer/FuzzerMain.cpp:20:10
    #13 0x7f3aff398bbc in __libc_start_main (/usr/grte/v4/lib64/libc.so.6+0x38bbc)
    #14 0x564718e525c8 in _start /usr/grte/v4/debug-src/glibc-2.19/csu/../sysdeps/x86_64/start.S:108

SUMMARY: AddressSanitizer: heap-use-after-free third_party/grpc/src/core/lib/gpr/mpscq.cc:49:43 in gpr_mpscq_pop_and_check_end(gpr_mpscq*, bool*)
Shadow bytes around the buggy address:
  0x0c527fff90f0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c527fff9100: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c527fff9110: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c527fff9120: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c527fff9130: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
=>0x0c527fff9140: fd fd fd fd[fd]fd fd fd fd fd fd fd fd fd fd fd
  0x0c527fff9150: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c527fff9160: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c527fff9170: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c527fff9180: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c527fff9190: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==498769==ABORTING
```

**Comment by yashykt** [comment]
The issue is that -
1) A timer is being set with a callback to &t->retry_initiate_ping_locked in maybe_initiate_ping. [source](https://github.com/grpc/grpc/blob/100e0a78ebc1ecfaefb57c5105506b3091d4b86c/src/core/ext/transport/chttp2/transport/writing.cc#L91)
2) A ref on transport is not being taken when we initialize the timer.
3) This means that if the timer fires after the transport is destroyed then it will result in a heap use-after-free resulting in possible SEGFAULTs

**Comment by yashykt** [comment]
To fix this, #14739 takes a ref before initializing the timer and unrefs it at the end of the function retry_initiate_ping_locked
